tensorflow_encrypted/session.py

```diff
@@ -0,0 +1,135 @@
+                    sanitized_fetches.append(fetch)
+
+        return sanitized_fetches
+
+    def decode_fetches(self, fetches: Any, fetches_out: Any) -> List[Union[tf.Tensor, tf.Operation]]:
+        if not isinstance(fetches, Iterable) or isinstance(fetches, tf.Tensor):
+            if not isinstance(fetches, tf.Tensor) and not isinstance(fetches, tf.Operation):
+                decoder = fetches
+                to_decode = decoder.value_on_0.from_decomposed(fetches_out)
+                return decoder.prot._decode(to_decode, decoder.is_scaled)
+            else:
+                return fetches_out
+
+        for idx, fetch in enumerate(fetches):
+            if isinstance(fetch, Iterable) and not isinstance(fetch, tf.Tensor):
+                for sub_idx, sub_fetch in enumerate(fetch):
+                    if not isinstance(sub_fetch, tf.Tensor) and not isinstance(sub_fetch, tf.Operation):
+                        decoder = fetches[idx][sub_idx]
+                        to_decode = decoder.value_on_0.from_decomposed(fetches_out[idx][sub_idx])
+                        fetches_out[idx][sub_idx] = decoder.prot._decode(to_decode, decoder.is_scaled)
+            else:
+                if not isinstance(fetch, tf.Tensor) and not isinstance(fetch, tf.Operation):
+                    decoder = fetches[idx]
+                    to_decode = decoder.value_on_0.from_decomposed(fetches_out[idx])
+                    fetches_out[idx] = decoder.prot._decode(to_decode, decoder.is_scaled)
+
+        return fetches_out
+
+    def run(
+        self,
+        fetches: Any,
+        feed_dict: Dict[str, np.ndarray] = {},
+        tag: Optional[str] = None,
+        write_trace: bool = False
+    ) -> Any:
+
+        sanitized_fetches = self.sanitize_fetches(fetches)
+
+        if not __TFE_STATS__ or tag is None:
+            fetches_out = self.sess.run(
+                sanitized_fetches,
+                feed_dict=feed_dict
+            )
+        else:
+            session_tag = "{}{}".format(tag, _run_counter[tag])
+            run_tag = os.path.join(__TENSORBOARD_DIR__, session_tag)
+            _run_counter[tag] += 1
+
+            writer = tf.summary.FileWriter(run_tag, self.sess.graph)
+            run_options = tf.RunOptions(trace_level=tf.RunOptions.FULL_TRACE)
+            run_metadata = tf.RunMetadata()
+
+            fetches_out = self.sess.run(
+                sanitized_fetches,
+                feed_dict=feed_dict,
+                options=run_options,
+                run_metadata=run_metadata
+            )
+
+            writer.add_run_metadata(run_metadata, session_tag)
+            writer.close()
+
+            if __TFE_TRACE__ or write_trace:
+                chrome_trace = timeline.Timeline(run_metadata.step_stats).generate_chrome_trace_format()
+                with open('{}/{}.ctr'.format(__TENSORBOARD_DIR__, session_tag), 'w') as f:
+                    f.write(chrome_trace)
+
+        return self.decode_fetches(fetches, fetches_out)
```

Comment:
you should be able to decode before fecthing

Reply:
Not sure why, but if I try to do that, some examples start to hang.